sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleCoalesceExec.scala

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util
+
+import ai.rapids.cudf.{HostMemoryBuffer, JCudfSerialization, NvtxColor, NvtxRange}
+import ai.rapids.cudf.JCudfSerialization.SerializedTableHeader
+
+import org.apache.spark.TaskContext
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.Partitioning
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Coalesces serialized tables on the host up to the target batch size before transferring
+ * the coalesced result to the GPU. This reduces the overhead of copying data to the GPU
+ * and also helps avoid holding onto the GPU semaphore while shuffle I/O is being performed.
+ * @note This should ALWAYS appear in the plan after a GPU shuffle when RAPIDS shuffle is
+ *       not being used.
+ */
+case class ShuffleCoalesceExec(child: SparkPlan, @transient rapidsConf: RapidsConf)
+    extends UnaryExecNode with GpuExec {
+
+  import GpuMetricNames._
+  override lazy val additionalMetrics: Map[String, SQLMetric] = Map(
+    NUM_INPUT_ROWS -> SQLMetrics.createMetric(sparkContext, DESCRIPTION_NUM_INPUT_ROWS),
+    NUM_INPUT_BATCHES -> SQLMetrics.createMetric(sparkContext, DESCRIPTION_NUM_INPUT_BATCHES),
+    "collectTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "collect batch time"),
+    "concatTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "concat batch time")
+  )
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    throw new IllegalStateException("ROW BASED PROCESSING IS NOT SUPPORTED")
+  }
+
+  override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    val metricsMap = metrics
+    // Keep the batch size below a 2GB limit so that no nested type child column can exceed
+    // the cudf 2^31 row count limit
+    val targetBatchByteSize = Math.min(rapidsConf.gpuTargetBatchSizeBytes, Integer.MAX_VALUE)
```

Comment:
seems like this protection should be on the config itself. Does it make sense for the target batch size to be above MAX_VALUE?

Reply:
Sure, I'll update the config's upper limit across the board.